feat: ComponentClient tracing

octonato opened the pull request:
Alternative to https://github.com/lightbend/kalix-jvm-sdk/pull/2046

octonato commented on sdk/java-sdk-spring/src/main/scala/kalix/javasdk/impl/reflection/Reflect.scala:
```diff
@@ -50,4 +54,32 @@ object Reflect {
 
   }
+
+  private implicit val stringArrayOrdering: Ordering[Array[String]] =
+    Ordering.fromLessThan(util.Arrays.compare[String](_, _) < 0)
+
+  implicit val methodOrdering: Ordering[Method] =
+    Ordering.by((m: Method) => (m.getName, m.getReturnType.getName, m.getParameterTypes.map(_.getName)))
```
Orthogonal to main change in this PR. 
I merged ReflectionUtils with Reflect class.

octonato commented on sdk/java-sdk-spring/src/main/scala/kalix/javasdk/impl/reflection/Reflect.scala:
```diff
@@ -50,4 +54,32 @@ object Reflect {
+  implicit val methodOrdering: Ordering[Method] =
+    Ordering.by((m: Method) => (m.getName, m.getReturnType.getName, m.getParameterTypes.map(_.getName)))
+
+  def lookupComponentClientField(instance: Any): List[ComponentClient] = {
```
This is the main utility method being added. 
For the moment in use when receiving Action commands, but not yet for Workflow

octonato commented on sdk/java-sdk-spring/src/main/scala/kalix/javasdk/impl/action/ReflectiveActionRouter.scala:
```diff
@@ -50,23 +51,30 @@ class ReflectiveActionRouter[A <: Action](
-          s"Couldn't find any method with input type [$inputTypeUrl] in Action [$action].")
-
+    // lookup ComponentClient
+    val maybeComponentClient = Reflect.lookupComponentClientField(action.getClass, action)
```
There is no test for this, but @efgpinto's [PR](https://github.com/lightbend/kalix-jvm-sdk/pull/2026/files) will add a sample that will then cover it.

efgpinto commented on sdk/java-sdk-spring/src/main/java/kalix/javasdk/client/ComponentClient.java:
```diff
@@ -40,10 +42,20 @@ public class ComponentClient {
     this.kalixClient = kalixClient;
   }
 
+  public void setCallMetadata(Metadata callMetadata) {
```
Should we document that this is intended for internal use? or is it ok for users to use this? if so, having this and the `.withMetadata` after building the deferred call might be confusing...

octonato replied:
Right. Actually this is a oversight of mine. I was planning to make it internal API. 

On the other hand, now I'm wondering what is the impact for using `withMetadata` in the `DeferredCall`. In both impls, we are setting the metadata, but we probably want to copy the tracing over, if any.

efgpinto reviewed (commented):
This LGTM with a small question. I will just try to run my sample against this to test before ✅ .

octonato commented:
@efgpinto, a few new things in this PR.

* made ComponentClient an interface and moved impl to ComponentClientImpl
* moved KalixClient to a `impl` package and undeprecated it
* integrated Cisco's PR (https://github.com/lightbend/kalix-jvm-sdk/pull/2022)

octonato commented:
Maybe good to review it commit by commit. 

First one is about setting the metadata on the client (bf5fe2e0e4f2ffe47b5ea67cae540638fc944719)
Second commit is integrating Cisco's PR (d4397b0d34f551536864bde0aba9680b8224569c)
Third is about moving the KalixClient to internal package (9f9d956ae4d322f071306a8087e8964fff17870a)

Last is just about formatting.

aludwiko commented on sdk/java-sdk-spring/src/main/java/kalix/javasdk/client/ActionCallBuilder.java:
```diff
@@ -16,194 +16,178 @@
-import akka.japi.function.Function7;
-import akka.japi.function.Function8;
-import akka.japi.function.Function9;
+import akka.japi.function.*;
```
I think we should avoid them.

octonato replied:
That was not intentional, I will revert it.

octonato replied:
Done

aludwiko commented on sdk/java-sdk-spring/src/main/scala/kalix/javasdk/impl/reflection/Reflect.scala:
```diff
@@ -50,4 +55,32 @@ object Reflect {
+  implicit val methodOrdering: Ordering[Method] =
+    Ordering.by((m: Method) => (m.getName, m.getReturnType.getName, m.getParameterTypes.map(_.getName)))
+
+  def lookupComponentClientField(instance: Any): List[ComponentClientImpl] = {
```
why is this a list of ComponentClients? I would expect only a single one or none.

octonato replied:
Me too, but we have not guarantee that how many component client a user will try to inject.

efgpinto commented on sdk/java-sdk-spring/src/main/scala/kalix/javasdk/impl/action/ReflectiveActionRouter.scala:
```diff
@@ -50,23 +51,33 @@ class ReflectiveActionRouter[A <: Action](
+    // lookup ComponentClient
+    val componentClients =
+      Reflect
+        .lookupComponentClientField(action.getClass, action)
```
Small bug found when trying this with the sample..

```suggestion
        .lookupComponentClientField(action)
```

efgpinto commented on sdk/java-sdk-spring/src/main/scala/kalix/javasdk/impl/action/ReflectiveActionRouter.scala:
```diff
@@ -76,21 +87,31 @@ class ReflectiveActionRouter[A <: Action](
+    // lookup ComponentClient
+    val componentClients =
+      Reflect
+        .lookupComponentClientField(action.getClass, action)
```
Small bug found when trying this with the sample..

```suggestion
        .lookupComponentClientField(action)
```

octonato replied:
I changed the signature from (Class, Any) to (Any), but how is it possible that the compiler didn't pick it?

efgpinto replied:
The other method receives `Any` and this was considered a `Tuple2`.. so it worked. I was just not sure if this was because of some implicit lying around or if it doesn't even need that.. 🤔

octonato replied:
Yeah, I was about to ask which error you saw. 
To be honest, I thought this kind of language quirk was already out of 2.13, but apparently not. 

Or we probably had a compiler warning, but didn't see because we have some many already.

By the way, I working on a branch that will make warnings fatal. Like we do in Akka and Kalix Runtime.

efgpinto reviewed (commented):
Partial review with small bug found and a typo

octonato commented:
We still had two tests using KalixClient directly. CI should pass now. 

I also changed the error message when someone tries to inject a 'illegal' dependency. Error was only showing the simple class name. Better to use a FQCN so users can easily find the offending class.

efgpinto commented:
Found another small issue.. The `traceparent` / `tracestate` don't seem to be injected automatically when executing (`execute()`) the deferred call.

octonato commented:
> Found another small issue.. The `traceparent` / `tracestate` don't seem to be injected automatically as headers when `.execute()`ing the deferred call.

@franciscolopezsancho, can you have a look at this? Seems to be something that was missing/broken in #2022.

franciscolopezsancho commented:
> Found another small issue.. The `traceparent` / `tracestate` don't seem to be injected automatically as headers when `.execute()`ing the deferred call.

@efgpinto Looking at the code seems fine but obviously you found something. Could you share your reproducer? if any?

franciscolopezsancho commented:
LGTM. I can create an example to test it all out if you'd like.
